### DISCUSSION TITLE
trashed pages never get purged

### DISCUSSION THREAD

**Discussion by bmm-alc** [outline/outline#3540] [General] [answered]
Hello,

The pages never get removed from trash. Is there any action to perform (like an environment variable to set, a toggle somewhere I've forgotten) ? 

![image](https://user-images.githubusercontent.com/80455946/168553254-f6f27e69-6675-48f6-9dde-9d6cd069d298.png)

outline version 0.63.0

best

**Comment by tommoor** [comment, marked as answer]
Needs documentation, but you should set up a cron job that calls `yourinstallation.com/api/utils.gc?token=<UTILS_SECRET>`, once a day is enough. `UTILS_SECRET` is whatever you define it as in the `.env`

**Reply by almereyda** [reply]
Ah, this Cron job is also new to me. Indeed it might be lacking from documentation. I'll figure a way to instantiate that within the Compose setup within the week, and report back then.

**Reply by tommoor** [reply]
Adding to the compose would be great, I can update the docs – should be as easy as a curl command and a tiny cron image I'd imagine

**Reply by almereyda** [reply]
We can either have cron installed to the Outline image, in so `cron -f` can be called from the `command:` line in foreground mode, or we go for a simple approximation such as presented in

- https://lab.libreho.st/libre.sh/compose/nextcloud/-/blob/2799eaef81ca31c8fab193d59ae3427130e2a615/docker-compose.yml#L66-73

needing to have `curl` present, too:

```
    entrypoint: |
      bash -c 'bash -s <<EOF
      trap "break;exit" SIGHUP SIGINT SIGTERM
      while /bin/true; do
        /usr/bin/curl -sL https://app.getoutline.com/api/utils.gc?token=${UTILS_SECRET}
        sleep 86400
      done
      EOF'

```

In a perfect world, we have the `entrypoint` of Outline changed to `yarn run '$@'` and provide a dedicated `command: start` to the container, which will be passed to the entrypoint, for plain use, and add a secondary `command: cron` to execute a custom cron scheduler through [Yarn's support for `package.json` run scripts](https://classic.yarnpkg.com/en/docs/package-json/#toc-scripts), eventually not needing to trigger it from outside, but inside the application: The whole environment will be the same between both containers, who use the same image (OverlayFS deduplicates this in RAM), but execute another command within it.

Which idea do you like most?

**Reply by tommoor** [reply]
I do not want to include cron inside the published docker container as it prevents portability. It needs to work across a variety of environments including cases where there are lots of server processes running 👍

**Reply by almereyda** [reply]
I see!

This is what I could come up with this evening:

```yaml
  cron:
    image: curlimages/curl:latest
    entrypoint: |
      /bin/ash -c '/bin/ash -s <<EOF
      trap "break;exit" SIGHUP SIGINT SIGTERM
      while /bin/true; do
        date +%Y-%m-%dT%H:%M:%S | xargs -I {} printf "%s " "{}"
        /usr/bin/curl -s http://outline:3000/api/utils.gc?token=${OUTLINE_UTILS_SECRET}
        printf "\n"
        sleep 86400
      done
      EOF'
    networks: ["outline"]
    logging:
      driver: "json-file"
      options:
        max-size: "10m"
```

This is an opinionated setup which differs from the one in the documentation in so far, as it:

- does not use host port mappings
- introduces a separate `"outline"` network for the application
  - which allows us to use Docker-internal DNS for service discovery
  - and directly reach Outline at `http://outline:3000` without needing the round trip and TLS overhead via a public interface.  
    Of course this can be changed to use an `https://` URL, which also allows to omit the extra network for such environments, where security standards are higher than ours here.
- Compose will replace the value for `${OUTLINE_UTILS_SECRET}` from `.env` for us, and we "do not even need" to add the secret to the container's environment. This is string interpolation at work, so the token will be visible, since the command is execute within an infitine loop.  
  You can inspect a running `cron` container of the above likes with:  
  `docker-compose exec cron ash -c "cat /proc/1/cmdline | xargs -0 echo"`
  - This could possibly be omitted with using a proper container task scheduler, such as `crontab` itself, where the `curl` command will only be executed at a certain interval, and the secret can be taken from the environment each time.
  - Either way the secret is part of the runtime environment of the `cron` container, one cleaner or dirtyier way, or the other.

Compose will report the log output like so (with different dates and similar times; this was me testing with `sleep 2`):

```log
cron_1      | 2022-05-16T19:54:42 {"success":true,"status":200,"ok":true}
cron_1      | 2022-05-16T19:54:44 {"success":true,"status":200,"ok":true}
cron_1      | 2022-05-16T19:54:46 {"success":true,"status":200,"ok":true}
cron_1      | 2022-05-16T19:54:48 {"success":true,"status":200,"ok":true}
cron_1      | 2022-05-16T19:54:50 {"success":true,"status":200,"ok":true}
```

While log aggregation receives this from the container:

```
2022-05-16T19:55:19 {"success":true,"status":200,"ok":true}
```

- [ ] If the endpoin'st response included a timestamp in the JSON response itself, we could omit the whole `date` boilerplate. I'd refrain from adding another `jq` dependency to get the timestamp into the JSON for now ;)

The opinionated setup was documented in https://github.com/outline/outline/discussions/3291#discussioncomment-2460649
